index.bs

```diff
@@ -3134,6 +3217,66 @@ and an optional [=boolean=] |isNoised| (default false):
+1. Return false.
+
+To <dfn>remove source and associated event reports</dfn> given an [=attribution source=] |source| and a [=moment=] |minTriggerTime|:
+1. [=set/iterate|For each=] [=event-level report=] |report| of the [=event-level report cache=]:
+    1. If |report|'s [=event-level report/source identifier=] is not equal to |source|'s [=attribution source/source identifier=], [=iteration/continue=].
+    1. If |report|'s [=event-level report/trigger time=] is less than |minTriggerTime|, [=iteration/continue=].
+    1. [=set/Remove=] |report| from the [=event-level report cache=].
+    1. [=list/Remove=] all elements from [=attribution rate-limit records=] whose [=attribution rate-limit record/entity ID=] is equal to |report|'s [=event-level report/report ID=].
+1. [=set/Remove=] |source| from the [=attribution source cache=].
+
+To <dfn>remove sources with attribution scopes not selected for destination</dfn> given a [=site=] |destination| and an [=attribution source=] |pendingSource|:
+1. Let |attributionScopeRecords| be a new [=list=].
+1. [=set/iterate|For each=] |attributionScope| in |pendingSource|'s [=attribution source/attribution scopes=]:
+    1. [=list/Append=] the [=tuple=] (|attributionScope|, |pendingSource|'s [=attribution source/source time=], |pendingSource|'s [=attribution source/source identifier=]) to |attributionScopeRecords|.
+1. [=set/iterate|For each=] |source| of the [=attribution source cache=]:
+    1. If |source|'s [=attribution source/reporting origin=] and |pendingSource|'s [=attribution source/reporting origin=] are not [=same origin=], [=iteration/continue=].
+    1. If |source|'s [=attribution source/attribution destinations=] does not [=set/contain=] |destination|, [=iteration/continue=].
+    1. [=set/iterate|For each=] |attributionScope| in |source|'s [=attribution source/attribution scopes=]:
+        1. [=list/Append=] the [=tuple=] (|attributionScope|, |source|'s [=attribution source/source time=], |source|'s [=attribution source/source identifier=]) to |attributionScopeRecords|.
+1. [=list/sort in ascending order|Sort=] |attributionScopeRecords| in ascending order with |a| being less than |b| if any of the following are true:
+      * |a|[1] is greater than |b|[1].
+      * |a|[1] is equal to |b|[1] and |a|[0] is greater than |b|[0].
+1. Let |selectedAttributionScopes| be a new [=set=].
+1. Let |sourcesToRemove| be a new [=set=].
+1. [=list/iterate|For each=] |record| of |attributionScopeRecords|:
+    1. If |selectedAttributionScopes|'s [=set/size=] is less than |pendingSource|'s [=attribution source/attribution scope limit=], [=set/append=] |record|[0] to |selectedAttributionScopes|.
+    1. Otherwise, if |selectedAttributionScopes| does not [=set/contain=] |record|[0], [=set/append=] |record|[2] to |sourcesToRemove|.
+1. [=set/iterate|For each=] |source| of the [=attribution source cache=]:
```

Comment:
It looks like `Remove source and associated event reports` only needs the source identifier, so we don't have to iterate the source cache again.